bee-common/bee-common/src/event.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! A module that provides a generic, type-safe event bus for arbitrary event types.
+
+use dashmap::DashMap;
+use std::any::{Any, TypeId};
+
+type Listener<'a> = dyn Fn(&dyn Any) + Send + Sync + 'a;
+
+/// An event bus for arbitrary event types.
+pub struct Bus<'a, ID = TypeId> {
+    listeners: DashMap<TypeId, Vec<(Box<Listener<'a>>, ID)>>,
```

Comment:
Why does ID default to TypeID ?

Reply:
I imagine that this will be the most common use-case.

Reply:
Yeah, I think I just have a hard time imagining it not being TypeID, did you have an example in mind ?

Reply:
Nothing specific. It just felt weird (as you mentioned in a comment before) to make the assumption that handlers would be tied to workers/specific types.

Reply:
Are we sure we want to use `DashMap`?

Reply:
Agree with Alex, maybe we should avoid using DashMap for now till the new release?

Reply:
To be fair, most of the locks we thought were due to Dashmap ended up doing something really different. To this day, I am not sure we ever had a lock due to Dashmap. The whole protocol uses thousands of operations a second and we had no lock in ages.

Reply:
Not saying we should absolutely keep it, but I am not sure we're having issues with it.